### PR TITLE
Add sys/storage/raft/bootstrap option to reset TLS keyring

### PR DESCRIPTION
**Pull request by Shaeli**
This PR resolves https://github.com/hashicorp/vault/issues/14032

It adds a new `reset_tls_keyring` optional parameter for the `sys/storage/raft/bootstrap` endpoint.

This will help recovering a vault cluster using raft for HA management only and not storage.

Currently, if a cluster is using raft for its HA management but another storage system (mysql, S3 etc), it's not possible to easily recover vault data when all servers of the cluster are not functioning anymore.

When re-creating vault nodes and connecting them to the storage, no node can be elected active because there is no raft cluster assembled to conduct the vote. Using the bootstrap endpoint (`sys/storage/raft/bootstrap`) is failing with :

```
$ vault write -f sys/storage/raft/bootstrap
Error writing data to sys/storage/raft/bootstrap: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/sys/storage/raft/bootstrap
Code: 500. Errors:

* could not generate TLS keyring during bootstrap: TLS keyring already present
```

There is a workaround to recreate a vault cluster, which is:
- Start vault in recovery mode
- Generate a recovery token
- Run `VAULT_TOKEN=<recovery-token> vault delete sys/raw/core/raft/tls`
- Restart vault in non-recovery mode
- Run `vault write -f sys/storage/raft/bootstrap` : the current node is now leader and no data was lost.

But this is not ideal as it's dangerously manipulating the storage. As the storage is persistent and not lost with the dead servers (storage is not RAFT), we should be able to re-bootstrap the cluster and recover from new servers.

The issue is also described here: https://github.com/hashicorp/vault/issues/14032 and mentioned here: https://discuss.hashicorp.com/t/how-to-migrate-from-non-ha-backend-to-non-ha-backend-raft-ha-backend/13702/10

I'm not familiar with golang, so please let me know if anything can be improved :)

### CODE REVIEW

**Review by drawks — APPROVED** [review]
This looks good to me. I'm wondering if there is anything else that we can do to get a maintainer to review this.

**Comment by drawks** [comment]
@hsimon-hashicorp any chance you can help solicit a reviewer on this? I'm here again running a very similar disaster recovery simulation as in https://github.com/hashicorp/vault/issues/14032 and noticing that this MR to fix/solve the issue has been sitting idle since it was submitted back in January. Any help would be appreciated.

File: http/sys_raft.go
Comment (VioletHynes) [inline http/sys_raft.go:41]: Instead of this new parameter, I think we should dynamically determine in `raftCreateTLSKeyring` whether or not we're using raft storage. This avoids the need for a new option, and all of the added code in this area

File: vault/raft.go
Comment (VioletHynes) [inline vault/raft.go:698]: Instead of this check, we could do something like this:
```
	if _, usingRaftStorage := c.underlyingPhysical.(*raft.RaftBackend); usingRaftStorage {
		`return nil, fmt.Errorf("TLS keyring already present")`
	}
```

This avoids the need to add a new parameter, and I think is a lot cleaner. The user shouldn't need to provide a boolean if we can figure it out ourselves

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:57]: This should have a GoDoc (comment beginning with the test name explaining what the test does above the test). It doesn't need to be long since you've already got a great description over at `testRaftHARecoverCluster`.

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:155]: nit: remove this t.Log

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:156]: nit: instead of this declaration, we could just pass 'nil' to the functions we're currently passing `conf` too

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:159]: Love this comment!

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:168]:
```suggestion
```

Comment (VioletHynes) [reply vault/external_tests/raftha/raft_ha_test.go:168]: These should no longer be needed due to improvements to `NewTestCluster`

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:200]: nit: I think this (and some of the comparisons below) would be a little cleaner if we used `require` here. e.g.:

```
	err := leaderClient.Sys().Mount("kv", &api.MountInput{
		Type: "kv-v2",
	})
        require.NoError(err)
```

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:232]: Please use camelCase here, i.e. `clusterRestored`

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:276]: We should split this into multiple lines to make sure this doesn't panic. For example:

```
data := secretRaw.Data["data"]
dataAsMap, ok := data.(map[string]interface{})
require.True(ok)
require.NotNil(dataAsMap)
require.Equal("awesome", dataAsMap[kittens])
```

**Review by VioletHynes — COMMENTED** [review]
Hi there! I'm very sorry this took such a long time to get eyes on. If you're still interested, I'm still happy to work towards getting this merged, as this would be a nice improvement to Vault.

I've outlined a few suggestions I'd like to see before I approve this. Most importantly, I think we can avoid the new option added to the `bootstrap` endpoint by checking to see if we're using raft storage instead. There's also some merge conflicts to resolve, but they seem simple enough.

Before getting this merged, we'd also need a changelog. This would be a file called `18615.txt` in the `changelog` directory that should look something like this (I'll leave the wording to you, feel free to look at the other files in that directory for inspiration):
release-note:bug
core: fixed issue when attempting to bootstrap HA when using Raft as HA but not storage

I love the test addition, and the quality of the submitted code is very encouraging 👌. If you have any questions about my comments, feel free to ask!

**Comment by drawks** [comment]
if @Shaeli isn't available/interested in updating this PR I can probably find some time to adopt it and update as needed as long as we've got organizational engagement from hashicorp that makes it likely to be re-reviewed with a high likelihood of merging. It is a bad game of tag if we keep the current pace around this issue.

**Comment by Shaeli** [comment]
👋 
@drawks I can update the PR with the suggestions, probably tomorrow or over the week end and 🤞 we can get this reviewed again and merged! 

Thanks @VioletHynes for the review

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [inline vault/external_tests/raftha/raft_ha_test.go:68]: You mentioned this in a comment, and actually I think it would be great if we could have a raft test here too, since nothing should be different.

This should be something like:
```
	t.Run("raft", func(t *testing.T) {
		physBundle := teststorage.MakeRaftBackend(t, logger, 0, nil, &raft.ClusterAddrBridge{
		    clusterAddressByNodeID: make(map[string]string),
	         })
		testRaftHARecoverCluster(t, physBundle, logger)
	})
```

**Review by VioletHynes — COMMENTED** [review]
This looks great! As you mentioned, I think it'd be great if we had raft as a test here too. Could you have a quick look to see if that would be easy enough to add?

I have a small comment on the wording of the comment that I'd like to see changed, but that's a minor thing.

Without the raft test, I'd still like to merge this, as I think the chance of issue is very low, but I would appreciate if you checked if one was easy to add. Let me know!

After you've looked into that, I'd be happy to merge this. Thanks for bearing with us for so long, and I appreciate the PR updates despite the wait.

File: vault/external_tests/raftha/raft_ha_test.go
Comment (VioletHynes) [reply vault/external_tests/raftha/raft_ha_test.go:69]: I appreciate the attempt, and thanks for reporting back. This should be covered by `TestRaft_HA_ExistingCluster` in `raft_ha_test.go`, so I'm not too worried, but it would have been nice to have the symmetry. Definitely nothing that'll block the PR. Thanks! I'm going to approve this and try and get this merged.

**Review by VioletHynes — APPROVED** [review]
Looks great! Thanks for the diligence here, as well as the patience in getting this reviewed and resolved.

I'm going to wait for all of the tests to run, but assuming no nasty surprises, I'll be merging this once we get back all green.

**Comment by Shaeli** [comment]
Seems like updating the branch with almost 2y changes broke the test, I'm checking !

**Comment by VioletHynes** [comment]
Right, yeah, looks like a few of the old references need to be updated, and you may need to find an alternative to `TestRaftServerAddressProvider`. I'm not sure why it was removed, but I don't think there'd be much harm in keeping whatever the body was as a local function for this test. There are some examples around the code of how to create these elsewhere, like `NewHardcodedServerAddressProvider`.

**Comment by Shaeli** [comment]
@VioletHynes This should be fixed 🤞

**Comment by VioletHynes** [comment]
Thanks for the quick update! Running the tests again 🤞

**Comment by Shaeli** [comment]
Do you know by chance if this test failure is something known / expected? 😬 
This is not related to my change
```
=== FAIL: vault Test_handleActivityWriteData/write_entities (0.01s)
    logical_system_activity_write_testonly_test.go:577: 
        	Error Trace:	/home/runner/work/vault/vault/vault/logical_system_activity_write_testonly_test.go:577
        	Error:      	"[]" should have 1 item(s), but has 0
        	Test:       	Test_handleActivityWriteData/write_entities
```

**Comment by VioletHynes** [comment]
That seems to me like a known flaky and your code shouldn't touch this at all. I'm giving it a re-run. I'm pretty sure we're all good on your end, but I'll keep re-running until we're green just in case.

**Comment by VioletHynes** [comment]
Thanks for your contribution, and again, apologies for how long it took to get eyes on this. This was a great PR and I appreciate your quick responses and high quality, tested code.

Merged :)

**Pull request merged**
